internal/evmreader/evmreader.go

```diff
@@ -0,0 +1,364 @@
+}
+
+// EvmReader reads inputs from the blockchain
+type EvmReader struct {
+	client      EthClient
+	wsClient    EthWsClient
+	inputSource InputSource
+	repository  EvmReaderRepository
+	config      NodePersistentConfig
+}
+
+func (r *EvmReader) String() string {
+	return "evm-reader"
+}
+
+// Creates a new EvmReader
+func NewEvmReader(
+	client EthClient,
+	wsClient EthWsClient,
+	inputSource InputSource,
+	repository EvmReaderRepository,
+	config NodePersistentConfig,
+) EvmReader {
+	return EvmReader{
+		client:      client,
+		wsClient:    wsClient,
+		inputSource: inputSource,
+		repository:  repository,
+		config:      config,
+	}
+}
+
+func (r *EvmReader) Run(
+	ctx context.Context,
+	ready chan<- struct{},
+) error {
+
+	for {
+		watchForNewInputsError := r.watchForNewBlocks(ctx, ready)
+		// If the error is a SubscriptionError restart watchForNewBlocks
+		// that will restart the subscription
+		if _, ok := watchForNewInputsError.(*SubscriptionError); !ok {
+			return watchForNewInputsError
+		}
+		slog.Error(watchForNewInputsError.Error())
+		slog.Info("Restarting subscription")
+	}
+}
+
+// Watch for new blocks and reads new inputs based on the
+// default block configuration, which have not been processed yet.
+func (r *EvmReader) watchForNewBlocks(
+	ctx Context,
+	ready chan<- struct{},
+) error {
+	headers := make(chan *Header)
+	sub, err := r.wsClient.SubscribeNewHead(ctx, headers)
+	if err != nil {
+		return fmt.Errorf("could not start subscription: %v", err)
+	}
+	slog.Info("Subscribed to new block events")
+	ready <- struct{}{}
+	defer sub.Unsubscribe()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-sub.Err():
+			return &SubscriptionError{Cause: err}
+		case <-headers:
+			// Every time a new block arrives
+			err = r.checkForNewInputs(ctx)
+			if err != nil {
+				slog.Error("Error checking got new inputs",
+					"error",
+					err,
+				)
+			}
+
+		}
+	}
+}
+
+// Check if is there new Inputs for all running Applications
+func (r *EvmReader) checkForNewInputs(ctx Context) error {
+
+	// Get All Applications
+	apps, err := r.repository.GetAllRunningApplications(ctx)
+	if err != nil {
+		return err
+	}
+
+	groupedApps := r.classifyApplicationsByLastProcessedInput(apps)
+
+	for lastProcessedBlock, apps := range groupedApps {
+
+		// Safeguard: Only check blocks starting from the block where the InputBox
+		// contract was deployed as Inputs can be added to that same block
+		if lastProcessedBlock < r.config.InputBoxDeploymentBlock {
+			lastProcessedBlock = r.config.InputBoxDeploymentBlock - 1
+		}
+
+		currentMostRecentFinalizedHeader, err := r.fetchMostRecentHeader(
+			ctx,
+			r.config.DefaultBlock,
+		)
+		if err != nil {
+			slog.Error("Error fetching most recent block",
+				"last default block",
+				r.config.DefaultBlock,
+				"error",
+				err)
+			continue
+		}
+		currentMostRecentFinalizedBlockNumber := currentMostRecentFinalizedHeader.Number.Uint64()
+
```

Comment:
Shouldn't this whole block be moved outside the loop for `groupedApps` (BTW, `appGroup` would be better) to retrieve `currentMostRecentFinalizedBlockNumber` only once and share it between apps?